runtime/src/global_request_scheduler_mts.c

```diff
@@ -0,0 +1,333 @@
+
+	int rc = priority_queue_enqueue_nolock(mgrq->sandbox_requests, sandbox_request);
+	if (rc == -ENOSPC) panic("Module's Request Queue is full\n");
+	// debuglog("Added a sandbox_request to the MGRQ");
+
+	/* Maintain the minheap structure by Removing and Adding the MGRQ from and to the global runqueue.
+	 * Do this only when the MGRQ's priority is updated.
+	 */
+	if (priority_queue_peek(mgrq->sandbox_requests) < last_mrq_deadline) {
+		priority_queue_delete_nolock(destination_queue, mgrq);
+
+		rc = priority_queue_enqueue_nolock(destination_queue, mgrq);
+		if (rc == -ENOSPC) panic("Global Runqueue is full!\n");
+		// debuglog("Added the MGRQ back to the Global runqueue - %s to Heapify", QUEUE_NAME);
+	}
+
+	LOCK_UNLOCK(&global_lock);
+
+	return sandbox_request;
+}
+
+/**
+ * @param pointer to the pointer that we want to set to the address of the removed sandbox request
+ * @returns 0 if successful, -ENOENT if empty
+ */
+int
+global_request_scheduler_mts_remove(struct sandbox_request **removed_sandbox_request)
+{
+	/* This function won't be used with the MTS scheduler. Keeping merely for the polymorhism. */
+	return -1;
+}
+
+
+/**
+ * @param removed_sandbox_request pointer to set to removed sandbox request
+ * @param target_deadline the deadline that the request must be earlier than to dequeue
+ * @returns 0 if successful, -ENOENT if empty or if request isn't earlier than target_deadline
+ */
+int
+global_request_scheduler_mts_remove_if_earlier(struct sandbox_request **removed_sandbox_request,
+                                                       uint64_t                 target_deadline)
+{
+	/* This function won't be used with the MTS scheduler. Keeping merely for the polymorhism. */
+	return -1;
+}
+
+/**
+ * @param removed_sandbox_request pointer to set to removed sandbox request
+ * @param target_deadline the deadline that the request must be earlier than to dequeue
+ * @param mt_class the multi-tenancy class of the global request to compare the target deadline against
+ * @returns 0 if successful, -ENOENT if empty or if request isn't earlier than target_deadline
+ */
+int
+global_request_scheduler_mts_remove_with_mt_class(struct sandbox_request **removed_sandbox_request,
+                                                          uint64_t                 target_deadline,
+                                                          enum MULTI_TENANCY_CLASS target_mt_class)
+{
+	/* Avoid unnessary locks when the target_deadline is tighter than the head of the Global runqueue */
+	uint64_t global_guaranteed_deadline = priority_queue_peek(global_request_scheduler_mts_guaranteed);
+	uint64_t global_default_deadline    = priority_queue_peek(global_request_scheduler_mts_default);
+
+	switch (target_mt_class) {
+	case MT_GUARANTEED:
+		if (global_guaranteed_deadline >= target_deadline) return -ENOENT;
+		break;
+	case MT_DEFAULT:
+		if (global_guaranteed_deadline == UINT64_MAX && global_default_deadline >= target_deadline)
```

Comment:
Yeah, this logic seems like it might benefit from having a function (i..e a name) associated with it as I'm seeing it again, and have to reverse engineer it each time.